### ISSUE TITLE
Fallback rewrites do not trigger after calling `notFound` function in dynamic route

### ISSUE THREAD

**Issue by jplwood** [vercel/next.js#71834]
### Link to the code that reproduces this issue

https://github.com/jplwood/nextjs-rewrite-notfound-repro

### To Reproduce

1. start the application with `npm run dev`

**Demonstrate the fallback rewrite:**

3. go to home route [https://localhost:3000/home](https://localhost:3000/home) 👉 page loads ✅ 
4. go to random route under home [https://localhost:3000/home/rewriteworks](https://localhost:3000/home/rewriteworks) 👉 rewrite to google is triggered ✅ 

**Demonstrate fallback not working**

5. go to dynamic route [https://localhost:3000/exists](https://localhost:3000/exists) 👉  page loads ✅ 
6. go to dynamic route again to trigger not found [https://localhost:3000/rewritesad](https://localhost:3000/rewritesad) 👉 page rewrite is not triggered, 404 page is shown ❌ 

### Current vs. Expected behavior

Per [rewrite docs](https://nextjs.org/docs/app/api-reference/next-config-js/rewrites):

> 6. fallback rewrites are checked/applied, these are applied before rendering the 404 page and after dynamic routes/all static assets have been checked.  [source](https://arc.net/l/quote/tvrnmbwd)

To me this means, if a route is not found, if there is a fallback rewrite, that should _always_ apply, no matter how that not found is initiated. i.e., 
- Implicit 404 ✅  - As shown in the first example, if the next router resolves a 404, the rewrite applies.
- explicit 404 ❌  - In the second example, if we're in a dynamic route, but, per whatever lookup logic we do (in my case, api call to a headless CMS), if the dynamic path isn't resolvable and you call [`notFound()`](https://nextjs.org/docs/app/api-reference/functions/not-found), this bypasses the fallback rewrite and goes straight to the 404 page.

Similarly, this seems like a topic to potentially mention in the [Dynamic Routes doc page](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes)?—how to handle dynamic routes that are not found (i.e., `notFound()`, `rewrites`, `redirects`, etc...)

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:39:46 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6020
  Available memory (MB): 32768
  Available CPU cores: 10
Binaries:
  Node: 20.10.0
  npm: 10.3.0
  Yarn: 1.22.22
  pnpm: 9.12.1
Relevant Packages:
  next: 15.0.2-canary.6 // Latest available version is detected (15.0.2-canary.6).
  eslint-config-next: N/A
  react: 19.0.0-rc-1631855f-20241023
  react-dom: 19.0.0-rc-1631855f-20241023
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Runtime

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local)

### Additional context

I tested my reproduction in Next 15.x and 13.x w/ App Router, both have the issue. I suspect this is just an app router issue.

**Comment by ztanner** [comment]
Hi,

Thank you for the report. I believe this is a misinterpretation of the docs and the behavior you're seeing is expected. The Next.js router is matching the `/rewritesad` request to the `/[slug]` page component. As described in the [matching order](https://nextjs.org/docs/app/building-your-application/routing/middleware#matching-paths) documentation, `fallback` rewrites are processed last, after dynamic routes. And since a dynamic route was matched, it doesn't continue on to try and match it against the fallback rewrite. 

The line quoted from the docs is referring to the fallback 404 handling that Next.js will do if it is unable to match a path after finishing the matching order described above, not for an explicit `notFound()` that you call from a component. 

If you'd like to specify a known list of paths that should be handled by your `[slug]` function while all other unknown paths don't match (and fallback to your rewrite), you can leverage `generateStaticParams` & `dynamicParams = false` ([docs](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams)), like so:

```tsx
export default function dynamic({ params }) {
  const { slug } = params
  if (slug === 'exists') {
    return <>hello world</>
  }

  return <div>...</div>
}

export function generateStaticParams() {
  return [{ slug: 'exists' }]
}

export const dynamicParams = false
```

**Comment by jplwood** [comment]
Thanks for explaining the misunderstanding @ztanner. After rereading, I guess this is a feature request then.

Just to clarify our use case. We have a cms-driven app that we render dynamically. The intention is to check to see if the route exists in the CMS:
    - if it does, render it
    - if it doesn't exist, rewrite to our legacy cms app
In theory we could use generateStaticParams, but we don't want to statically render all of the pages from our CMS (related to a current hosting limitation). Our current workaround is to check a cache of known routes in the new CMS via Middleware, and rewrite if it isn't a known path. This is a little clunky and makes all requests a little heavier than they'd need to be if we could trigger a rewrite when a "matched" dynamic route, ultimately, when checked, isn't found.

In other words, ideally calls to `notFound()` would check rewrite rules before showing the 404 page.
